packages/embark/src/lib/core/config.js

```diff
@@ -122,15 +122,16 @@ Config.prototype.reloadConfig = function() {
 };
 
 Config.prototype.loadContractFiles = function() {
-  const contracts = this.embarkConfig.contracts;
-  const newContractsFiles = this.loadFiles(contracts);
-  if (!this.contractFiles || newContractsFiles.length !== this.contractFiles.length || !deepEqual(newContractsFiles, this.contractFiles)) {
-    this.contractsFiles = this.contractsFiles.concat(newContractsFiles).filter((file, index, arr) => {
-      return !arr.some((file2, index2) => {
-        return file.path === file2.path && index < index2;
-      });
-    });
-  }
+  const loadedContractFiles = this.loadFiles(this.embarkConfig.contracts);
+  // `this.contractsFiles` could've been mutated at runtime using
+  // either `config:contractsFiles:add` event or through calls to
+  // `loadExternalContractsFiles()`, so we have to make sure we preserve
+  // those added files before we reset `this.contractsFiles`.
+  //
+  // We do that by determining the difference between `loadedContractFiles` and the ones
+  // already in memory in `this.contractsFiles`.
+  const addedContractFiles = this.contractsFiles.filter(existingFile => !loadedContractFiles.some(file => file.originalPath === existingFile.originalPath));
+  this.contractsFiles = loadedContractFiles.concat(addedContractFiles);
```

Comment:
Tested this with contract files being added via plugins. `config:contractsFiles:add` works too as it also just adds entries to `contractsFiles` at run time.

We'll check for differences between `loadedContractFiles` and `contractsFiles` either way.